### DISCUSSION TITLE
Unreachable error when calling vips_init() in Cloudflare Workers environment (WASM)

### DISCUSSION THREAD

**Discussion by mojodevagentsix** [libvips/libvips#4216] [General]
I'm encountering an issue when trying to use libvips compiled to WebAssembly within a Cloudflare Workers environment. Specifically, the vips_init() function is throwing an "unreachable" error.

Environment:

Cloudflare Workers
libvips compiled to WebAssembly using Emscripten

Steps to Reproduce:

1. Compile libvips to WebAssembly using Emscripten with the flags listed below.
2. Load the resulting .wasm file in a Cloudflare Worker.
3. Attempt to call the vips_init() function.

Expected Behavior: The vips_init() function should initialize libvips successfully.

Actual Behavior: An "unreachable" error is thrown when calling vips_init().

Error Message:

`Error in vips_init: RuntimeError: unreachable
    at vips-es6.wasm.vips_init (vips-es6.wasm-01392cc6:0x12c496)
    at Object.fetch (index.js:165:47)
    at async jsonError (.internal-6d243f3f-ce0c-4f2f-bbd2-40a2a2c4db25-facade-1.js:12:12)`

Compilation Flags:

`-sAUTOLOAD_DYLIBS=0
-sABORTING_MALLOC=0
-sSTANDALONE_WASM
-sERROR_ON_UNDEFINED_SYMBOLS=0
-sMODULARIZE
-sEXPORT_NAME='Vips'
-sINITIAL_MEMORY=134217728
-sMAXIMUM_MEMORY=134217728
-sSTACK_SIZE=8MB
-gsource-map
-sALLOW_TABLE_GROWTH
-sALLOW_MEMORY_GROWTH=0
-sTEXTDECODER=2
-sASSERTIONS=2
-sFORCE_FILESYSTEM=0
-sMALLOC='emmalloc'
-sUSE_OFFSET_CONVERTER
-sVERBOSE=1
-sDEMANGLE_SUPPORT=1
-sEXPORTED_FUNCTIONS=_get_vips_version
--no-entry
-sEXCEPTION_STACK_TRACES
-sBINARYEN_EXTRA_PASSES=--emit-target-features
--profiling
--profiling-funcs
--keep-names
--emit-symbol-map
--threadprofiler
-sSAFE_HEAP=1
-sCHECK_NULL_WRITES=0
-sLIBRARY_DEBUG
-sEXCEPTION_DEBUG
-sSYSCALL_DEBUG
-sDYLINK_DEBUG
-sSTACK_OVERFLOW_CHECK=2
-sRUNTIME_DEBUG=1
-Os  # Optimize for size
-g  # Include debug information`

Additional Context:

- The same WebAssembly module works correctly when used in a browser environment with index.html or React.
- The native C version of the code works as expected.
- The issue only occurs in the Cloudflare Workers environment.

Questions

1. Are there known issues with running libvips in a Cloudflare Workers environment?
2. Are there specific memory or environment settings required for libvips to initialize properly in this context?
3.  Could this be related to the memory limitations or execution environment of Cloudflare Workers?
4. Are there any flags in our compilation settings that might be causing conflicts in the Workers environment?
5. Is the fixed memory allocation (-sALLOW_MEMORY_GROWTH=0) a potential issue in the Workers context?

Observations:

1. I am  using a fixed memory allocation (134217728 bytes, or 128MB) without growth.
2. The stack size is set to 8MB(tested with 256kb, 2MB, 4MB but same error), which might be larger than necessary for a Worker environment.
3. I am  using emmalloc as the memory allocator, which is optimized for size but might have different behavior than the default allocator.
4. Many debug flags are enabled, which could impact performance but should help with diagnosing the issue.

Any insights or suggestions on how to resolve this "unreachable" error when initializing libvips in a Cloudflare Workers environment, particularly considering our compilation settings, would be greatly appreciated.

**Comment by kleisauke** [comment]
Please see https://github.com/kleisauke/wasm-vips/issues/2.

**Reply by mojodevagentsix** [reply]
@kleisauke : 
 I was also getting the memory bound errors, but I changed the compilation flags for debuging , then getting these errors for various function, when I try to call vips_init() 
 
 `index.js:158 Error calling vips_init: RuntimeError: unreachable
    at vips-es6.wasm (vips-es6.wasm-0132fd26:0x3066e6)
    at vips-es6.wasm.attempt_allocate (vips-es6.wasm-0132fd26:0x2f2675)
    at vips-es6.wasm.emmalloc_memalign (vips-es6.wasm-0132fd26:0x2f24d9)
    at vips-es6.wasm.emmalloc_malloc (vips-es6.wasm-0132fd26:0x2f283e)
    at vips-es6.wasm.g_mutex_impl_new (vips-es6.wasm-0132fd26:0x188b15)
    at vips-es6.wasm.g_mutex_lock_impl (vips-es6.wasm-0132fd26:0x188b7e)
    at vips-es6.wasm.g_mutex_lock (vips-es6.wasm-0132fd26:0x17bffa)
    at vips-es6.wasm.g_logv (vips-es6.wasm-0132fd26:0x173e83)
    at vips-es6.wasm.g_log (vips-es6.wasm-0132fd26:0x173a1c)
    at vips-es6.wasm.vips_init (vips-es6.wasm-0132fd26:0x1237a8)`
    
    Please share your knowledge for this, this is in cloudflare worker env only.

**Reply by kleisauke** [reply]
Cloudflare Workers does not currently support threading nor the Web Workers API. If you require these features, you'll need to reach out to Cloudflare about this lack of support if you need to use their services.

**Reply by mojodevagentsix** [reply]
Hi @kleisauke ,
I am trying the same on my local machine but getting the same error not rechable , so Do I have to run specific function before running vips_init(), 
Now I am trying to achieve the same in local environment with node.js and express server.


`Visit http://localhost:3000/init-vips to initialize the Vips module
WASM module loaded successfully
Available exports: [
  '__getTypeName',
  '_embind_initialize_bindings',
  '__wasm_call_ctors',
  'vips_init',
  'get_vips_version',
  'strerror',
  '__indirect_function_table',
  'free',
  'malloc',
  'ntohs',
  '_emscripten_tls_init',
  'pthread_self',
  '_initialize',
  '_emscripten_thread_init',
  '_emscripten_thread_crashed',
  'emscripten_main_thread_process_queued_calls',
  'htons',
  'htonl',
  'emscripten_main_runtime_thread_id',
  'emscripten_stack_get_base',
  'emscripten_stack_get_end',
  '_emscripten_run_on_main_thread_js',
  '_emscripten_thread_free_data',
  '_emscripten_thread_exit',
  'emscripten_get_sbrk_ptr',
  'sbrk',
  '_emscripten_check_mailbox',
  'setThrew',
  '_emscripten_tempret_set',
  'emscripten_stack_init',
  'emscripten_stack_set_limits',
  'emscripten_stack_get_free',
  '_emscripten_stack_restore',
  '_emscripten_stack_alloc',
  'emscripten_stack_get_current',
  '__cxa_demangle',
  '__cxa_free_exception',
  '__cxa_increment_exception_refcount',
  '__cxa_decrement_exception_refcount',
  '__get_exception_message',
  '__cxa_can_catch',
  '__cxa_is_pointer_type',
  '__set_stack_limits'
]
Calling __wasm_call_ctors
_emscripten_init_main_thread_js called
Error during WebAssembly instantiation: RuntimeError: unreachable
    at wasm://wasm/010ec636:wasm-function[7568]:0x2d9813
    at wasm://wasm/010ec636:wasm-function[7831]:0x2e25a5
    at wasm://wasm/010ec636:wasm-function[7504]:0x2d7398
    at wasm://wasm/010ec636:wasm-function[59]:0x510a
    at safeCall (file:///Users/rahul/projects/nodeprojects/wasm-node-express-project/server.js:18:29)
    at loadWasmModule (file:///Users/rahul/projects/nodeprojects/wasm-node-express-project/server.js:191:5)
    at async file:///Users/rahul/projects/nodeprojects/wasm-node-express-project/server.js:210:27`
    
    
    Do I have to run specific function in order to initiliaze the vips_init ? 
    
    
    
    
    
vips-es6.js , vips-es6.wasm works fine with index.html, or react but when I am trying to load the vips0es6.wasm in node the wasm module is loading but the moment I am trying to call vips_init I am having error unreachable, 

Do I have to call the following 

__wasm_call_ctors'
_emscripten_tls_init'
  '_initialize'
  
  Please share your thoughs for this

**Reply by kleisauke** [reply]
I would start by removing these debug and instrumentation flags:
```
-sASSERTIONS=2 --threadprofiler -sSAFE_HEAP=1 -sCHECK_NULL_WRITES=0 -sLIBRARY_DEBUG -sEXCEPTION_DEBUG -sSYSCALL_DEBUG -sDYLINK_DEBUG -sSTACK_OVERFLOW_CHECK=2 -sRUNTIME_DEBUG=1
```

Additionally, note that `-sSTANDALONE_WASM` is incompatible with Emscripten's current pthread integration, as thread creation requires JavaScript support.

**Reply by mojodevagentsix** [reply]
Hello @kleisauke,

Following your suggestion, I have removed the debug flag and added -sUSE_PTHREADS=0. In Node.js, i am creating multiple workers, each loading the Vips module, which subsequently creates threads.

Here is the C++ function _generate_small_image, which is responsible for generating the small image:

Worker 7 Status:

Worker 7 is ready.
Vips WASM module initialized.
Vips instance within worker: Object [ 'incrementExceptionRefcount', 'decrementExceptionRefcount', 'getExceptionMessage', 'BindingError', 'InternalError', 'count_emval_handles', 'getInheritedInstanceCount', 'getLiveInheritedInstances', 'setAutoDeleteLater', 'flushPendingDeletes', 'setDelayFunction', '_generate_small_image', '_malloc', '_free', '_free_result', '__initialize', 'bigintToI53Checked', 'ENV', 'ccall', 'cwrap', 'addFunction', 'UTF8ToString', 'stringToNewUTF8', 'FS', 'allocateUTF8', 'deletionQueue', 'HEAP8', 'HEAP16', 'HEAPU8', 'HEAPU16', 'HEAP32', 'HEAPU32', 'HEAPF32', 'HEAPF64', 'HEAP64', 'HEAPU64', 'calledRun', 'stdin', 'stdout', 'stderr' ]
Thread "Thread 1" (0x00000001) is now running.
However, when attempting to call generate_small_image via a POST method in Node.js, I encountered the following error


`About to call generate_small_image
generate_small_image completed
Raw result: Error: VIPS error - glib: Error creating thread: Resource temporarily unavailable
glib: Error creating thread: Resource temporarily unavailable
`



And here is my CPP code just for the reference , Its very basic 
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <vips/vips.h>


extern "C" {

 char* generate_small_image() {
    char* result = nullptr;
    VipsImage* image = nullptr;
    void* buffer = nullptr;
    size_t buffer_size = 0;
    char* base64 = nullptr;

    if (VIPS_INIT("generate_small_image")) {
        return strdup("Error: VIPS initialization failed");
    }

    // Create a simple colored image (red)
    unsigned char* data = (unsigned char*)malloc(200 * 100 * 3);
    if (!data) {
        result = strdup("Error: Failed to allocate memory for image data");
        goto cleanup;
    }
    for (int i = 0; i < 200 * 100 * 3; i += 3) {
        data[i] = 255;  // Red
        data[i+1] = 0;  // Green
        data[i+2] = 0;  // Blue
    }

    image = vips_image_new_from_memory(data, 200 * 100 * 3, 200, 100, 3, VIPS_FORMAT_UCHAR);
    if (!image) {
        result = strdup("Error: Failed to create image");
        goto cleanup;
    }

    // Write the image to a buffer as PNG
    if (vips_pngsave_buffer(image, &buffer, &buffer_size, nullptr)) {
        result = strdup("Error: Failed to save image to buffer");
        goto cleanup;
    }

    // Encode the buffer as base64
    base64 = g_base64_encode((const guchar*)buffer, buffer_size);
    if (!base64) {
        result = strdup("Error: Failed to encode image to base64");
        goto cleanup;
    }

    // Create the result string
    result = (char*)malloc(strlen(base64) + 50); // Extra space for JSON formatting
    if (!result) {
        result = strdup("Error: Memory allocation failed for result");
        goto cleanup;
    }

    sprintf(result, "{\"image\":\"%s\"}", base64);

cleanup:
    if (data) free(data);
    if (image) g_object_unref(image);
    if (buffer) g_free(buffer);
    if (base64) g_free(base64);
    
    const char* vips_error = vips_error_buffer();
    if (vips_error && *vips_error) {
        char* error_result = (char*)malloc(strlen(vips_error) + 50);
        if (error_result) {
            sprintf(error_result, "Error: VIPS error - %s", vips_error);
            if (result) free(result);
            result = error_result;
        }
    }
    
    vips_thread_shutdown();
    vips_error_clear();

    return result ? result : strdup("Error: Unknown error occurred");
}

 void free_result(char* ptr) {
    free(ptr);
}

}



This works fine in index.html react (browser) , but with node.js it's not though while compiling I do have vips-node.js , vips-node.wasm files.


If you have any insights on how to resolve this thread creation issue or if additional adjustments are needed, your guidance would be greatly appreciated. Thank you!

**Reply by mojodevagentsix** [reply]
My basic need is , I have a C++ code and which process the images, I have compiled the code to WASM and it works fine with react but my main purpose is I want to create an API which will load the wasm module and that exported function, and that function I can use in (cloudflare worker or from any other framework 
)

**Reply by kleisauke** [reply]
> [...] and added -sUSE_PTHREADS=0

Please remove this flag. As I said in the linked issue, wasm-vips requires threading functionality to be available, so it must be compiled and linked with the `-pthread` flag to work correctly.
